### DISCUSSION TITLE
[Feature] Why the service name must be unique under one env other than application？

### DISCUSSION THREAD

**Discussion by Camel9527** [kubevela/kubevela#1895] [Q&A]
I have severral applications with many services for each aplication. Got error when there are the same service name even if they are not under the same applicaion.

**Comment by wonderflow** [comment]
Yes, this is a known issue as we use the service name to create underlying resources. These resources should have unique name.

App myapp -- (service: mysvc) -- K8s Deployment(mysvc)

**Comment by cbl315** [comment]
It's the limitation of `kubernetes`, the name of k8s resources like `Deployment` is an attribute of its `metadata`.

Docs ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/

```
Each object in your cluster has a Name that is unique for that type of resource
```

**Comment by Camel9527** [comment]
@charles-woshicai 
why not add a app-name prefix before the service and deployment or other resources under app while they are created by kubevela?
For ex:
app name: demo-app1
service name: service1

then the service name in k8s is `demo-app1-service1`

**Comment by Camel9527** [comment]
@wonderflow 
So, it's not an `enhancement` in the nearest release but a `question` need to be discussed？

**Comment by wonderflow** [comment]
Yes, we need to discuss more. `a app-name prefix` doesn't fix the issue, it makes the underlying resource has complex name and can't be understood clearly.

**Comment by Camel9527** [comment]
@wonderflow 
There are many apps develop by different engineer running in one k8s cluster. Name conflict  is bothered.
Is there any recommend solution for this?

**Comment by Camel9527** [comment]
@wonderflow 
Also I cant specify the namespace of the application except for change env. But if specify different env for each applicaton, also need create all the workloads and traits under every env.

**Comment by wonderflow** [comment]
Will it be confuse for users to see their underlying component resources has a prefix of app-name? And was it possible for the user add the prefix by themselves?

**Comment by cbl315** [comment]
How about add a random hash to the end of name? Just like what pod was doing.

We could add `owner-reference` to bind the service and application.

E.g:
```
app name: demo-app1
service name: service1

runtime service name: service1-a8e809a
```

**Comment by cbl315** [comment]
> How about add a random hash to the end of name? Just like what pod was doing.
> 
> We could add `owner-reference` to bind the service and application.
> 
> E.g:
> 
> ```
> app name: demo-app1
> service name: service1
> 
> runtime service name: service1-a8e809a
> ```

And I think it could be an opitional feature, only the application with specific annotaion like `vela.io/randomName: 'enable'` would enable it

**Comment by Camel9527** [comment]
> How about add a random hash to the end of name? Just like what pod was doing.
> 
> We could add `owner-reference` to bind the service and application.
> 
> E.g:
> 
> ```
> app name: demo-app1
> service name: service1
> 
> runtime service name: service1-a8e809a
> ```

It's more confuse for users, cos I can't  konw which ones belong to my app except get more info by `kubectl describe` for ex.

**Comment by wonderflow** [comment]
> > How about add a random hash to the end of name? Just like what pod was doing.
> > We could add `owner-reference` to bind the service and application.
> > E.g:
> > ```
> > app name: demo-app1
> > service name: service1
> > 
> > runtime service name: service1-a8e809a
> > ```
> 
> It's more confuse for users, cos I can't konw which ones belong to my app except get more info by `kubectl describe` for ex.

yes, I won't vote for the random suffix design

**Comment by kingcrunch** [comment]
Hi,

That is a kind of a show stopper. Because now, if I don't want to maintain excel sheets with given service names per Namespace, the only solution is to provide one namespace per application, which can be massive with microservices. 

With `<appName>-<name>` one can at least minimize the issue, because `appName` is unique per namespace. This is how Helm does it and nobody I know was confused.

Why is it impossible to configure it myself?

This works

```
template: {
  outputs: {
    foo: {
      metadata: name: context.appName + "-" + context.name
    }
  }
} 
```

This does not work


```
template: {
  output: {
     metadata: name: context.appName + "-" + context.name
  }
} 
```

**Reply by wonderflow** [reply]
```
template: {
  output: {
     metadata: name: context.appName + "-" + context.name
  }
} 
```

What's the error reported? I think it could work

**Reply by kingcrunch** [reply]
That'd be great. Reading the documentation (and this discussion) I came to the conclusion, that it might be unsupported, but if I'm wrong, that would already solve my issue

There is no error. It just doesn't work

Vela-Core Image: `oamdev/vela-core:v1.1.0`

example.cue
```
example: {
    description: ""
    type: "component"
    attributes: workload: definition: {
        apiVersion: "apps/v1"
        kind:       "Deployment"
    }
    annotations: {}
    labels: {}
}

template: {
    output: {
        apiVersion: "apps/v1"
        kind:       "Deployment"
        metadata: {
            name: context.appName + "-" + context.name
            labels: {
                "app.kubernetes.io/name": context.appName
                "app.kubernetes.io/instance": context.appName
                "app.kubernetes.io/component": context.name
                "app.oam.dev/appRevision": context.appRevision
            }
        }
        // snip
    }
}
```

`application.yaml`

```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: foo
spec:
  components:
    - name: app
      type: example
```

```bash
$ kubectl -n example get deployments,pods
NAME                  READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/app   1/1     1            1           6m55s

NAME                       READY   STATUS    RESTARTS   AGE
pod/app-78b7df69cf-psv8k   1/1     Running   0          6m55s
```

With `metadat: name: context.appName + "-" + context.name` I would expect a deployment name `foo-app`.
